File: 백준/Gold/2504. 괄호의 값/괄호의 값.py
import sys
input = sys.stdin.readline

stack = []
brcket = list(map(str, input().rstrip()))

problem = 1
answer = 0
flag = True
for i in range(len(brcket)):
    if brcket[i] == "(":
        flag = True
        stack.append(brcket[i])
        problem *= 2
    elif brcket[i] == "[":
        flag = True
        stack.append(brcket[i])
        problem *= 3
    elif brcket[i] == ")":
        if len(stack) == 0 or stack[-1] != "(":
            print(0)
            exit()
        else:
            stack.pop()
            if not flag:
                problem //= 2
                continue
            else:
                answer += problem
                flag = False
                problem //= 2
    elif brcket[i] == "]":
        if len(stack) == 0 or stack[-1] != "[":
            print(0)
            exit()
        else:
            stack.pop()
            if not flag:
                problem //= 3
                continue
            else:
                answer += problem
                flag = False
                problem //= 3

if len(stack) != 0 or len(stack) != 0:
    print(0)
else:
    print(answer)

File: 백준/Gold/2504. 괄호의 값/괄호의 값.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;

		String line = br.readLine();
		ArrayList<Character> blank = new ArrayList<>();

		for (int i = 0; i < line.length(); i++) {
			blank.add(line.charAt(i));
		}

		solve(blank);

	}

	private static void solve(ArrayList<Character> blank) {
		Stack<Character> stack = new Stack<>();
		int answer = 0;
		int result = 1;

		for (int i = 0; i < blank.size(); i++) {
			Character c = blank.get(i);
			if (c == '(') {
				stack.push(c);
				result *= 2;
			} else if (c == '[') {
				stack.push(c);
				result *= 3;
			} else {

				if (c == ')') {
					if (stack.isEmpty() || stack.peek() != '(') {
						answer = 0;
						break;
					}

					if (blank.get(i-1) == '(') {
						answer += result;
					}

					result /= 2;
				}

				if (c == ']') {
					if (stack.isEmpty() || stack.peek() != '[') {
						answer = 0;
						break;
					}

					if (blank.get(i-1) == '[') {
						answer += result;
					}

					result /= 3;
				}
				stack.pop();
			}
		}

		if (!stack.isEmpty()) {
			answer = 0;
		}

		System.out.println(answer);
	}
}
